### PR TITLE
Estimate segment field usages

### PR DESCRIPTION
**Pull request by dnhatn**
We have introduced a new memory estimation method in serverless, based on the number of segments and the fields within them. This new approach works well overall, but it still falls short in cases where most fields are used more than once - for example, in both doc_values and postings, or doc_values and points. This change exposes the total usage of fields in segments, allowing us to adjust the memory estimate for these cases.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/codec/FieldInfosWithUsages.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    public static int computeUsages(FieldInfo[] infos) {
+        int usages = 0;
+        for (FieldInfo fi : infos) {
+            if (fi.getIndexOptions() != IndexOptions.NONE) {
+                usages++;
+            }
+            if (fi.hasNorms()) {
+                usages++;
+            }
+            if (fi.hasVectors()) {
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/index/codec/FieldInfosWithUsages.java:33]: We can probably skip term vectors: their memory usage does not scale with the number of fields (like stored fields).

**Review by jpountz — APPROVED** [review]

**Comment by jpountz** [comment]
The change looks fine. I wonder if we need to go this granular though, or if we should assume that all fields that exist have an index (either terms, points or vectors) and doc values. Some fields may only have doc values, but then it's fine if we overestimate a bit?

File: server/src/main/java/org/elasticsearch/index/codec/FieldInfosWithUsages.java (hunk 1 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/codec/FieldInfosWithUsages.java]: Yes, I have pushed https://github.com/elastic/elasticsearch/pull/112760/commits/5cc4f3524ad624d713dcc21cb0093cc252845d72 to remove this.

**Comment by dnhatn** [comment]
Thanks, Adrien. I considered this option, but it would require overestimating the current estimate by 20% in all cases, which might prevent us from running 2GB instances. I'll merge this PR and discuss the follow-up changes for serverless. If we find a different solution, I'll revert this change.

**Pull request merged**